enterprise/ha/src/main/java/org/neo4j/kernel/ha/SlaveUpdatePuller.java

```diff
@@ -177,25 +206,30 @@ public void run()
     }
 
     @Override
-    public void init() throws Throwable
+    public synchronized void init() throws Throwable
     {
-        // TODO Don't do this. This is just to satisfy LockSupport park/unpark
-        // And we cannot have this class extend Thread since there's a naming clash with Lifecycle for stop()
-        me = new Thread( this, UPDATE_PULLER_THREAD_PREFIX + instanceId );
-        me.start();
+        if ( shutdownLatch != null )
+        {
+            return; // This SlaveUpdatePuller has already been initialised
+        }
+
+        shutdownLatch = new BinaryLatch();
+        jobScheduler.schedule( JobScheduler.Group.pullUpdates, this );
     }
 
     @Override
-    public void shutdown() throws Throwable
+    public synchronized void shutdown() throws Throwable
     {
-        halted = true;
-        while ( me.getState() != Thread.State.TERMINATED )
+        if ( shutdownLatch == null )
         {
-            Thread.sleep( 1 );
-            Thread.yield();
+            return; // This SlaveUpdatePuller has already been shut down
         }
-        invalidEpochHandler = null;
-        me = null;
+
+        Thread thread = updatePullingThread;
```

Comment:
I need to read the `updatePullingThread` variable before I set `halted` to `true`. Otherwise, I'll be racing with the background thread, who can terminate at any time and write `null` to the `updatePullingThread`. I don't want to worry about that when I `unpark` it.